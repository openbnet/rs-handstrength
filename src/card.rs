#[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
pub enum Suit {
    S,
    H,
    C,
    D,
    A, // any *
    E, // TrainerSuit first suit in cards
    F, //2nd suit
    G, // 3rd suit
    X, // Except E - 3 possible
    Y, // Except E and F - 2 possible
    Z, // Except E, F, and G -4th suit

}

#[derive(Clone, Copy, Debug, Eq, Hash,PartialOrd, PartialEq, Ord)]
pub struct Card {
    pub value: u8,
    pub suit: Suit,
}
use std::fmt;

impl fmt::Display for Suit {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let suit = match self {
            Suit::C => "c",
            Suit::D => "d",
            Suit::H => "h",
            Suit::S => "s",
            Suit::A => "*",
            Suit::E => "e",
            Suit::F => "f",
            Suit::G => "g",
            Suit::X => "x",
            Suit::Y => "y",
            Suit::Z => "z",
        };
        write!(f, "{}", suit)
    }
}

impl fmt::Display for Card {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{} of {}", self.value, self.suit)
    }
}
impl Default for Card {
    fn default() -> Self {
        Card {
            value: 0,
            suit: Suit::S,
        }
    }
}