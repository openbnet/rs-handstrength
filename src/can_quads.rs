use crate::card::{Card, Suit};
use crate::sorting::get_same_value_map;
use crate::can_libs::*;
pub fn can_have_quads(cards: &[Card]) -> CanHaveCombis {
    let grouped_cards = get_same_value_map(cards);
    let mut quads_combinations: CanHaveCombis = Vec::new();

    for group in grouped_cards {
        let card_value = group[0].value;
        match group.len() {
            3 => {
                // Trips on the board
                let mut hand_matchers = Vec::new();
                for val in (2..=14).rev() {
                    if val != card_value {
                        hand_matchers.push(
                            vec![vec![
                            Card { value: card_value, suit: Suit::A },
                            Card { value: if val == 14 { 1 } else { val }, suit: Suit::A }]
                        ]);
                    }
                }
                println!("hand_matchers {:?}", hand_matchers);
                quads_combinations.push((
                    group,
                    hand_matchers 
                ));
            }
            2 => {
                // One pair on the board
                quads_combinations.push((
                    group,
                    vec![vec![vec![
                    Card { value: card_value, suit: Suit::A },
                    Card { value: card_value, suit: Suit::A }
                ]]]));
            }
            _ => {}
        }

    }

    quads_combinations
}
