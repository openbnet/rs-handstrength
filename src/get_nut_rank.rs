
use std::collections::HashMap;
use std::sync::Mutex;

use crate::can_strflush::*;
use crate::can_quads::*;
use crate::can_fullhouse::*;
use crate::can_flush::*;
use crate::can_str::*;
use crate::can_trips::*;
use crate::can_twopair::*;
use crate::can_pair::*;
use crate::can_highcard::*;
use crate::can_libs::*;
use crate::card::{Card, Suit};
use lazy_static::lazy_static;

lazy_static! {
    static ref CACHE: Mutex<HashMap<(Vec<Card>, Vec<Card>), (u16, NutRankType)>> = Mutex::new(HashMap::new());
}
#[derive(Debug,Clone, PartialEq, Eq)]
pub enum NutRankType {
    StraightFlush,
    Quads,
    FullHouse,
    Flush,
    Straight,
    Trips,
    TwoPair,
    Pair,
    HighCard
}
pub fn get_nut_rank(hand: &Vec<Card>, comcards: &Vec<Card>, relative: bool) -> (u16, NutRankType) {
    // get_nut_rank_uncached(hand, comcards, relative)
    let mut cache = CACHE.lock().unwrap();
    let key = (hand.clone(), comcards.clone());
    if let Some((rank, nut_rank_type)) = cache.get(&key) {
        return (*rank, nut_rank_type.clone());
    }
    let (rank, nut_rank_type) = get_nut_rank_uncached(hand, comcards, relative);
    cache.insert((hand.clone(), comcards.clone()), (rank, nut_rank_type.clone()));
    (rank, nut_rank_type)
}
pub fn get_nut_rank_uncached(hand: &Vec<Card>, comcards: &Vec<Card>, relative: bool) -> (u16, NutRankType) {
    let can_str_flush = can_have_straight_flush(comcards);
    // println!("can_str_flush {:?}", can_str_flush);
    let (hit, rank) = is_subset(hand, can_str_flush, relative); 
    // println!("str flush hit {:} rank {:}", hit, rank);
    if hit {
        return (rank, NutRankType::StraightFlush);
    }
    let mut score = rank.clone();
    let can_quads = can_have_quads(comcards);
    let (hitq, rankq) = is_subset(hand, can_quads, relative); 
    // println!("quads  {:} rank {:}", hitq, rankq);
    score += rankq; 
    if hitq {
        return (score, NutRankType::Quads);
    }
    

    let can_fh = can_have_fullhouse(comcards);
    let (hitfh, rankfh) = is_subset(hand, can_fh, relative); 
    // println!("can_fh  {:} rank {:}", hitfh, rankfh);
    score += rankfh; 
    
    if hitfh {
        return (score, NutRankType::FullHouse);
    }

    let can_fl = can_have_flush(comcards);
    let (hitfl, rankfl) = is_subset(hand, can_fl, relative); 
    // println!("can_fl  {:} rank {:}", hitfl, rankfl);
    score += rankfl; 
    if hitfl {
        return (score, NutRankType::Flush);
    }


    let can_str = can_have_str(comcards);
    let (hitstr, rankstr) = is_subset(hand, can_str, relative); 
    // println!("can_str  {:} rank {:}", hitstr, rankstr);
    score += rankstr; 
    if hitstr {
        return (score, NutRankType::Straight);
    }

    let can_trips = can_have_trips(comcards);
    let (hittrips, ranktrips) = is_subset(hand, can_trips, relative); 
    // println!("can_trips  {:} rank {:}", hittrips, ranktrips);
    score += ranktrips; 
    if hittrips {
        return (score, NutRankType::Trips);
    }

    let can_2pair = can_have_twopairs(comcards);
    let (hit2p, rank2p) = is_subset(hand, can_2pair, relative); 
    // println!("can_2pair  {:} rank {:}", hit2p, rank2p);
    score += rank2p; 
    if hit2p {
        return (score, NutRankType::TwoPair);
    }


    let can_pair = can_have_pair(comcards);
    let (hitp, rankp) = is_subset(hand, can_pair, relative); 
    // println!("can_pair  {:} rank {:}", hitp, rankp);
    score += rankp; 
    if hitp {
        return (score, NutRankType::Pair);
    }

    let can_hc = can_have_highcard(comcards);
    let (hithc, rankhc) = is_subset(hand, can_hc, relative); 
    // println!("can_hc  {:} rank {:}", hithc, rankhc);
    score += rankhc; 

    return (score, NutRankType::HighCard);
    
    
}
pub fn is_subset(hand: &Vec<Card>, combis: CanHaveCombis, relative: bool) -> (bool, u16) {
    let mut rank = 0;

    for (board, same_rank_hands) in combis {
        if relative {
            // println!("is subset relative board {:?} {:?}", board, hand);
            // [Card { value: 13, suit: C }, Card { value: 6, suit: C }, Card { value: 5, suit: C }] 
            // [Card { value: 13, suit: H }, Card { value: 8, suit: C }, Card { value: 7, suit: C }, Card { value: 10, suit: C }]
            // we need to see if any of the board cards are in hand
            // if there is any single match, we need to ignore this whole board combi if relative is on
            // else proceed as normal
            // however the tricky thing is to handle the suit::A case for the board
            // if its the board's suit::A, it should be ignored if the num of card value in hand + num of card value on board > 4

            
            let mut should_pass = true;
            for bc in &board {
                if bc.suit == Suit::A {
                  let num_in_board = board.iter().filter(|&card| card.value == bc.value).count();
                  let num_in_hand = hand.iter().filter(|&card| card.value == bc.value).count();
                    if num_in_board + num_in_hand > 4 {
                        should_pass = false;
                        break;
                    }  
                } else {
                    if hand.contains(&bc) {
                        should_pass = false;
                        break;
                    }
                }

            }

            if !should_pass {
                continue;
            }


        }
        for same_rank in same_rank_hands {
            if relative {
                let filtered_same_rank: Vec<Vec<Card>> = same_rank.iter()
                    .filter(|&handmatcher| {
                        for hc in handmatcher {
                            if hc.suit == Suit::A {
                                let num_in_board = board.iter().filter(|&card| card.value == hc.value).count();
                                let num_in_hand = hand.iter().filter(|&card| card.value == hc.value).count();
                                if num_in_board + num_in_hand > 4 {
                                    return false;
                                }  
                            } else {
                                if hand.contains(&hc) {
                                    return false;
                                }
                            }
                        }
                        return true;
                    })
                    .cloned()
                    .collect();
                let (hit, score) = matches_hand(hand, &filtered_same_rank);
                // println!("matches hand hit {:?} score {:?} rank {:?}", hit, score, rank);
                rank += score;
                // println!("rank {:?}", rank);
                if hit {
                    return (true, rank);
                }
            } else {
                let (hit, score) = matches_hand(hand, &same_rank);
                // println!("matches hand hit {:?} score {:?} rank {:?}", hit, score, rank);
                rank += score;
                // println!("rank {:?}", rank);
                if hit {
                    return (true, rank);
                }
            }
        }
    }
    (false, rank)
}

fn matches_hand(hand: &Vec<Card>, same_rank: &Vec<CardHand>) -> (bool, u16) {
    let mut score: u16 = 0;
    for hand_matcher in same_rank {
        if is_hand_match(hand, hand_matcher) {
            return (true, score);
        }
        score += 1;
    }
    (false, score)
}
// Helper function to determine if a hand matches a specific hand matcher
// matches any and one
fn is_hand_match(hand: &Vec<Card>, hand_matcher: &CardHand) -> bool {
    let mut matched_cards = 0;
    let mut r_hand = hand.clone();
    for matcher_card in hand_matcher {
        if let Some(index) = r_hand.iter().position(|card| matches_card(card, matcher_card)) {
            matched_cards += 1;
            // Remove the matched card from the remaining hand
            r_hand.remove(index);
        } else {
            return false;
        }
    }

    matched_cards == hand_matcher.len()
}

// Helper function to check if a single card matches a matcher card
fn matches_card(card: &Card, matcher_card: &Card) -> bool {
    (matcher_card.suit == Suit::A || card.suit == matcher_card.suit) && card.value == matcher_card.value
}

// fn get_cards_left_matchers(index: usize, hand: [Card;4], pEval)