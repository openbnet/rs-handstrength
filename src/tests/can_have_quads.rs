
#[cfg(test)]

mod tests {
    use crate::card::{Card, Suit};
    use crate::can_quads::can_have_quads;
    use insta::assert_debug_snapshot;
    use crate::get_nut_rank::*;
    #[test]
    fn quads_with_trips_board() {
        let cards = vec![
            Card { value: 10, suit: Suit::S },
            Card { value: 10, suit: Suit::H },
            Card { value: 10, suit: Suit::D },
            Card { value: 8, suit: Suit::S },
            Card { value: 5, suit: Suit::C },
        ];

        let quads_combinations = can_have_quads(&cards);
        // println!("quads_combinations, {:?}", quads_combinations);
        // Check for 13 possible combinations (excluding 10)
        assert_eq!(quads_combinations.len(), 1);
        // println!("first board {:?}", quads_combinations[0].1);
        assert_debug_snapshot!(quads_combinations);
        let hand = vec![
            Card { value: 10, suit: Suit::H },
            Card { value: 11, suit: Suit::D },
            Card { value: 2, suit: Suit::S },
            Card { value: 2, suit: Suit::C },
        ];
        let nr = get_nut_rank(&hand, &cards);
        // println!("nr {:?}", nr);
        assert_eq!(nr.0, 3);
    }
    #[test]
    fn quads_with_no_pair_board() {
        let cards = vec![
            Card { value: 11, suit: Suit::S },
            Card { value: 10, suit: Suit::H },
            Card { value: 8, suit: Suit::D },
            Card { value: 3, suit: Suit::S },
            Card { value: 5, suit: Suit::C },
        ];

        let quads_combinations = can_have_quads(&cards);
        assert_eq!(quads_combinations.len(), 0, "Should be no quads combinations");
    }
    #[test]
    fn quads_with_one_pair_board() {
        let cards = vec![
            Card { value: 10, suit: Suit::S },
            Card { value: 10, suit: Suit::H },
            Card { value: 8, suit: Suit::D },
            Card { value: 3, suit: Suit::S },
            Card { value: 5, suit: Suit::C },
        ];
    
        let quads_combinations = can_have_quads(&cards);
        assert_eq!(quads_combinations.len(), 1, "Should be one quads combination");
        // println!("quads_combinations {:?}", quads_combinations[0].1);
        assert_eq!(quads_combinations[0].1.len(), 1, "Combination one hand");
        assert_debug_snapshot!(quads_combinations)
    }
    #[test]
    fn quads_with_two_pair_board() {
        let cards = vec![
            Card { value: 10, suit: Suit::S },
            Card { value: 10, suit: Suit::H },
            Card { value: 8, suit: Suit::D },
            Card { value: 8, suit: Suit::S },
            Card { value: 5, suit: Suit::C },
        ];
    
        let quads_combinations = can_have_quads(&cards);
        assert_eq!(quads_combinations.len(), 2, "Should be two quads combinations");
        // println!("quads_combinations {:?}", quads_combinations);
        assert_debug_snapshot!(quads_combinations)
    }
     
}
