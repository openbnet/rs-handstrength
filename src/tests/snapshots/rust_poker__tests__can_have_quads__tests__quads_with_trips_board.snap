---
source: src/tests/can_have_quads.rs
expression: quads_combinations
---
[
    (
        [
            Card {
                value: 10,
                suit: S,
            },
            Card {
                value: 10,
                suit: H,
            },
            Card {
                value: 10,
                suit: D,
            },
        ],
        [
            [
                [
                    Card {
                        value: 10,
                        suit: A,
                    },
                    Card {
                        value: 1,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 10,
                        suit: A,
                    },
                    Card {
                        value: 13,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 10,
                        suit: A,
                    },
                    Card {
                        value: 12,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 10,
                        suit: A,
                    },
                    Card {
                        value: 11,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 10,
                        suit: A,
                    },
                    Card {
                        value: 9,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 10,
                        suit: A,
                    },
                    Card {
                        value: 8,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 10,
                        suit: A,
                    },
                    Card {
                        value: 7,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 10,
                        suit: A,
                    },
                    Card {
                        value: 6,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 10,
                        suit: A,
                    },
                    Card {
                        value: 5,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 10,
                        suit: A,
                    },
                    Card {
                        value: 4,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 10,
                        suit: A,
                    },
                    Card {
                        value: 3,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 10,
                        suit: A,
                    },
                    Card {
                        value: 2,
                        suit: A,
                    },
                ],
            ],
        ],
    ),
]
